controllers/core/cninode_controller.go

```diff
@@ -0,0 +1,101 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/aws/amazon-vpc-resource-controller-k8s/apis/vpcresources/v1alpha1"
+	rcHealthz "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/healthz"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/k8s"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/node/manager"
+	"github.com/go-logr/logr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type CNINodeController struct {
+	Log         logr.Logger
+	NodeManager manager.Manager
+	Context     context.Context
+	K8sAPI      k8s.K8sWrapper
+}
+
+func (s *CNINodeController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cniNode, err := s.K8sAPI.GetCNINode(req.Name, req.Namespace)
+	if err != nil {
+		s.Log.Error(err, "Failed to get CNINode")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if len(cniNode.Spec.Features) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	nodeName := cniNode.Name
+	s.Log.Info("The CNINode has been updated", "CNINode", nodeName, "Features", cniNode.Spec.Features)
+
+	if node, err := s.K8sAPI.GetNode(nodeName); err != nil {
+		// if not found the node, we don't requeue and just wait VPC CNI to send another request
+		s.Log.Info("CNINode reconciler didn't find the node and will do exponential retry", "Node", node.Name)
+		return ctrl.Result{}, err
+	} else {
+		// make sure we have the node in cache already
+		if _, found := s.NodeManager.GetNode(nodeName); !found {
+			s.Log.Info("CNINode controller could not find node, will try again", "NodeName", nodeName)
+			return ctrl.Result{Requeue: true}, nil
+		} else {
+			// add the node into working queue as UPDATE event
+			s.Log.Info("CNINode controller found the node and sending the node update")
+			if err = s.NodeManager.UpdateNode(nodeName); err != nil {
```

Comment:
As far as I can tell, the only purpose of this controller is to trigger UpdateNode events? I'm wondering if it would make sense to instead setup a Watch for CNINode for the Node reconciler.

Reply:
I don't think node controller has a good place to add a watch for this CRD since node controller is purely for watching node updates and operating on them. The independent controller for this CRD will operate on node object as a resource regarding the CRD updates and I want to set it up as an overlay mechanism for even more resources use cases.